toffy/streak_detection.py

```diff
@@ -0,0 +1,410 @@
+import numpy as np
+import os
+from typing import Union, Optional
+from pathlib import Path
+from ark.utils import load_utils
+from skimage import (
+    filters,
+    exposure,
+    restoration,
+    measure,
+    draw,
+    util,
+    io,
+    img_as_int,
+)
+from dataclasses import dataclass
+import pandas as pd
+from functools import partial
+
+
+@dataclass
+class StreakData:
+    """Contains data for correcting the streaks consisting of binary masks, dataframes with
+    location and size properties, a directory for saving, and the shape / channel for mask
+    generation. In addition provides a function to save any of the binary masks or dataframes.
+
+    Args:
+        shape (tuple): The shape of the image / fov.
+        streak_channel (str): The specific channel name used to create the masks.
+        corrected_dir (Path): The directory used to save the corrected tiffs and data in.
+        streak_mask (np.ndarray): The first binary mask indicating candidate streaks.
+        streak_df (pd.DataFrame): A dataframe, containing the location, area, and eccentricity
+        of each streak.
+        filtered_streak_mask (np.ndarray): A binary mask with out the false streaks.
+        filtered_streak_df (pd.DataFrame): A subset of the `streak_df` containing location, area
+        and eccentricity values of the filtered streaks.
+        boxed_streaks (np.ndarray): An optional binary mask containing an outline for each
+        filtered streaks.
+        corrected_streak_mask (np.ndarray): An optional binary mask containing the lines used for
+        correcting the streaks.
+    """
+
+    shape: tuple = None
+    streak_channel: str = None
+    corrected_dir: Path = None
+    streak_mask: np.ndarray = None
+    streak_df: pd.DataFrame = None
+    filtered_streak_mask: np.ndarray = None
+    filtered_streak_df: pd.DataFrame = None
+    boxed_streaks: np.ndarray = None
+    corrected_streak_mask: np.ndarray = None
+
+    def __getitem__(self, item):
+        return getattr(self, item)
+
+    def save_data(self, name: Optional[str] = None, all_data: bool = True):
```

Comment:
I think it makes more sense to move the saving functions out of the DataClass and have them as separate functions. Storing the intermediate files in the dataclass works well, but I think having the I/O stuff be separate will help to make this more modular.